File: wix-petri-server/src/it/java/com/wixpress/petri/petri/JdbcSpecsDaoIT.java
package com.wixpress.petri.petri;

import com.wixpress.petri.experiments.domain.ExperimentSpec;
import org.joda.time.DateTime;
import org.junit.Before;
import org.junit.Test;
import org.springframework.dao.DuplicateKeyException;

import java.util.List;

import static com.wixpress.petri.petri.SpecDefinition.ExperimentSpecBuilder.anExperimentSpec;
import static java.util.Arrays.asList;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.empty;
import static org.hamcrest.core.Is.is;

/**
 * Petri - (c) Wix LTD. http://www.wix.com
 */
public class JdbcSpecsDaoIT extends PetriDaoIT<ExperimentSpec, ExperimentSpec> {

    private ExperimentSpec experimentSpec;
    private ExperimentSpec updatedSpec;
    private JdbcSpecsDao deleteEnablingDao;

    @Before
    public void aSetup() {
        deleteEnablingDao = new JdbcSpecsDao(dbDriver.jdbcTemplate, new SpecMapper(objectMapper, mappingErrorHandler));
        dao = deleteEnablingDao;

        experimentSpec = anExperimentSpec("f.q.n", new DateTime()).withTestGroups(asList("on", "off")).build();
        updatedSpec = anExperimentSpec("f.q.n", new DateTime()).withTestGroups(asList("a", "b", "c")).build();
    }


    @Test
    public void insert() {
        ExperimentSpec inserted = dao.add(experimentSpec);
        assertThat(inserted, is(experimentSpec));
        List<ExperimentSpec> fetched = dao.fetch();
        assertThat(fetched, is(asList(experimentSpec)));
    }

    @Test(expected = DuplicateKeyException.class)
    public void insertDuplicateClassNameThrowsException() {
        dao.add(experimentSpec);
        dao.add(experimentSpec);
    }

    @Test
    public void update() {
        dao.add(experimentSpec);
        dao.update(updatedSpec, new DateTime());
        assertThat(dao.fetch(), is(asList(updatedSpec)));
    }

    @Test
    public void delete() {
        dao.add(experimentSpec);
        deleteEnablingDao.delete(experimentSpec.getKey());
        assertThat(dao.fetch(), is(empty()));
    }


    @Override
    protected void insertIllegalJson() {
        dbDriver.insertSpec("illegalSpec", "IRRELEVANT");
    }

    @Override
    protected ExperimentSpec objectToAdd() {
        return experimentSpec;
    }


}

File: petri-server-core/src/it/java/com/wixpress/petri/petri/JdbcSpecsDaoIT.scala
package com.wixpress.petri.petri

import java.io.IOException
import java.util.Arrays._

import com.wixpress.common.specs2.JMock
import com.wixpress.petri.experiments.jackson.ObjectMapperFactory
import com.wixpress.petri.petri.SpecDefinition.ExperimentSpecBuilder._
import org.joda.time.DateTime
import org.specs2.mutable.{Before, SpecWithJUnit}
import org.springframework.dao.DuplicateKeyException

import scala.collection.JavaConverters._

/**
 * @author dmitryk
 * @since 21-Sep-2015
 */
class JdbcSpecsDaoIT extends SpecWithJUnit with JMock {

  sequential

  "JdbcScalaSpecsDao" should {

    "return empty list" in new ctx {
      dao.fetch() must beEmpty
    }

    "return null on non-deserializable experiment" in new ctx {
      checking {
        oneOf(mappingErrorHandler).handleError(`with`(contain("illegalSpec")), `with`(contain("experiment")), `with`(any[IOException]))
      }

      dbDriver.insertSpec("illegalSpec", "IRRELEVANT")

      dao.fetch() must be_==(Seq(null))
    }

    "insert" in new ctx {
      val inserted = dao.add(experimentSpec)
      inserted must be_==(experimentSpec)
      dao.fetch() must be_==(Seq(experimentSpec))
    }

    "throw on duplicate key" in new ctx {
      dao.add(experimentSpec)
      dao.add(experimentSpec) must throwA[DuplicateKeyException]
    }

    "update" in new ctx {
      dao.add(experimentSpec)
      dao.update(updatedSpec, nowTime)
      dao.fetch() must be_==(Seq(updatedSpec))
    }

    "delete" in new ctx {
      dao.add(experimentSpec)
      dao.delete(experimentSpec.getKey)
      dao.fetch() must beEmpty
    }

  }

  private def nowTime = DateTime.now()

  trait ctx extends Before {
    val now = nowTime

    val dbDriver = DBDriver.dbDriver(DBDriver.JDBC_H2_IN_MEM_CONNECTION_STRING)
    val objectMapper = ObjectMapperFactory.makeObjectMapper
    val mappingErrorHandler = mock[MappingErrorHandler]
    val dao = new JdbcSpecsDao(dbDriver.jdbcTemplate, new SpecMapper(objectMapper, mappingErrorHandler))

    val experimentSpec = anExperimentSpec("f.q.n", now).withTestGroups(asList("on", "off")).build
    val updatedSpec = anExperimentSpec("f.q.n", now).withTestGroups(asList("a", "b", "c")).build

    override def before: Any = {
      dbDriver.createSchema()
    }
  }

}
